### ISSUE TITLE
Composite types/generics

### ISSUE THREAD

**Issue by glhrmfrts** [luna/luna#70]
I just noticed that the parser currently accepts only identifiers as a hash literal key, e.g.:
```
{hello: "world"}
```
I think that when this was implemented it was meant to be temporary, but since Luna is planned to be a  statically typed language I think we need to define a syntax for structures such as a hash that uses generic types.

**Comment by vendethiel** [comment]
please, use always-matching brackets, not c++-style `c++`

**Comment by glhrmfrts** [comment]
@vendethiel sorry, but what do you mean?

**Comment by vendethiel** [comment]
Sorry, i meant `<>`, which are not always matched... Making parsing harder.

**Comment by glhrmfrts** [comment]
Yeah I don't see much sense in that being unmatched.
Anyway, this syntax is very ugly to me:
```
map<string, int>{
  "number": 2
}
```
But i can't think of anything else.

**Comment by vendethiel** [comment]
"Being unmatched" => they are when you use it as comparison operators.

`map[string, int]{...}`
`map!(string, int){...}` (d-style)

**Comment by aisk** [comment]
I think if luna have a type inference, the syntax will be cool:

```
{"foo": "bar"}    => map<string, string>
{"foo": "bar", "alist": [1, 2, 3]}    => map<string, list<any>>
{"foo": "bar", "truth": 42}    => map<string, any>
```

But I think this will be a large project 😅

**Comment by vendethiel** [comment]
very, very large indeed. And can be very surprising – see Scala.

**Comment by aisk** [comment]
@vendethiel Yes, may be this is not luna's target `simple, elegant, explicit ` haha

**Comment by glhrmfrts** [comment]
@vendethiel Oh I see now :p 
But maybe `<>` would be easier than square brackets (because of subscript), unless it's context-dependent and we do some semantic analysis later.

**Comment by glhrmfrts** [comment]
@aisk Maybe it's possible to a large project achieve simplicity? lol

**Comment by aisk** [comment]
@glhrmfrts I think it's not just too complex to implement, but with this design luna's type system may be more complex, and make it more hard too learn and use.

**Comment by glhrmfrts** [comment]
Indeed, a more strict type system might be good for both sides.
